torch/_functorch/aot_autograd.py

```diff
@@ -2036,6 +2095,298 @@ def wrapped_compiled_fn(args):
+                        curr_outs_unwrapped += unwrapped_tensors
+                        unwrap_infos.append(SubclassCreationMeta(
+                            creation_lambda=lambda tensors, fn=subclass_type.__tensor_unflatten__, ctx=ctx: fn(tensors, ctx),
+                            flat_tensor_start_idx=unwrapped_outs_count,
+                            arg_count=len(unwrapped_tensors),
+                        ))
+                        unwrapped_outs_count += len(unwrapped_tensors)
+                    else:
+                        curr_outs_unwrapped.append(a)
+                        unwrap_infos.append(unwrapped_outs_count)
+                        unwrapped_outs_count += 1
+
+                if store_output_meta:
+                    # This is very shady (I don't like it, but can't think of a better alternative)
+                    # At runtime, we need access to metadata that tells us how to map
+                    # the dense output tensors back into subclass tensors that the user expects
+                    # We need this info separately for both the fw graph and bw graph
+                    # The only way to actually get this info is by running the fw and bw at tracetime.
+                    #
+                    # I don't want to have to trace the entire function AGAIN (and increase the compile time),
+                    # So instead I rely here on this metadata getting populated
+                    # at the time that we make_fx() the final traced function.
+                    #
+                    # Alternatively we could run functionalization and make_fx() directly here,
+                    # although keeping them separate feels cleaner.
+                    nonlocal subclass_meta
+                    if is_primals:
+                        assert subclass_meta.fw_output_infos is None
+                        subclass_meta.fw_output_infos = unwrap_infos
+                        assert subclass_meta.fw_output_types is None
+                        subclass_meta.fw_output_types = [type(x) for x in outs]
+                    else:
+                        assert subclass_meta.bw_output_infos is None
+                        subclass_meta.bw_output_infos = unwrap_infos
+                return curr_outs_unwrapped
+            if is_joint:
+                assert len(outs_wrapped) == 2
+                assert isinstance(outs_wrapped[0], (list, tuple))
+                assert isinstance(outs_wrapped[1], (list, tuple))
+                outs_unwrapped = (
+                    type(outs_wrapped[0])(unwrap_output_list(outs_wrapped[0], is_primals=True)),
+                    type(outs_wrapped[1])(unwrap_output_list(outs_wrapped[1], is_primals=False)),
+                )
+            else:
+                assert isinstance(outs_wrapped, list)
+                outs_unwrapped = list(unwrap_output_list(outs_wrapped, is_primals=True))
+            return outs_unwrapped
+        return wrapped_flat_fn
```

Comment:
High level comment, without a concrete direction for how to fix: this is too long. Let's figure out how to make it shorter.